---
title: "Data Visualization"
author: "Ricardo gore"
date: "Thuesday 7/31/2018"
output: html_document
# output_dir: "docs"
---

```{r echo = FALSE, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Please, read my outlines for the data visualization course in the the follow README file [here](https://github.com/RJEGR/July_2018_bioinfo) and deep in the knowlege of this advantage.

## Preface

When I started coding for biology I realize on this amazing challenge about how to tell the history from a bunch of Next Generation sequencing datasets. Visualization of information (from massive data mining in special) become in a nice part of my data scientist training.


## Trainning dataset

We will use high throughput sequencing dataset from a marine non model organism exposed to hidrocarbon polutant. The libraries sequenced were acqured from undifferented (Un) and sexual differented stages ( Male and female). In this experiment time 0 and three corresponded to hidrocarbon pollutant expossition (before and after, respectivily).

Libraries were preprocessing using the standar parameters within trimmomatic and the assembly were performed with trinity. Differential gene expression and annotation were perform with edgeR (an R package) and Trinotate, respectivily.

The chunk-code here correspond to my workflow used in the lab. Hope you enjoy it!

## Required package:

```{r}
# library(devtools)
# devtools::install_github("rlbarter/superheat")
# install_github("cstubben/trinotateR")
# devtools::install_github("ropensci/taxa")
# devtools::install_github("grunwaldlab/metacoder")
# if (!require("DT")) install.packages('DT')
# source("https://bioconductor.org/biocLite.R"); biocLite("org.Hs.eg.db"); (~ 74.3 MB)

```

## Loading input files

First lets 

```{r echo = TRUE}
dir <- c("/Users/cigom/Documents/GitHub/July_2018_bioinfo/infile/")
setwd(dir)
# x <- read.table(paste0(dir,"diffExpr.P1e-2_C1.matrix"))
```

And read the input file; then let's make a `head(x)` from the file:

```{r echo = FALSE, results='asis'}
x <- read.table(paste0(dir,"diffExpr.P1e-2_C1.matrix2"))
knitr::kable(head(x))
```

Let's log2 transform the data
```{r data, echo = TRUE}
data = x # restore before doing various data transformations
data = log2(data+1)
data = as.matrix(data) # convert to matrix
data = t(scale(t(data), scale=F)) # Centering rows
```


### Heatmap from Differential gene Expression
Using [superheat](https://rlbarter.github.io/superheat/index.html)

```{r, fig.cap = "example 1", fig.align = "center", echo=TRUE}
library(superheat)
superheat(data,
          # retain original order of rows/cols
          pretty.order.rows = TRUE,
          pretty.order.cols = TRUE,
          row.dendrogram = TRUE,
          left.label = "none",
          bottom.label.text.angle = 0,
          row.title = "Differential Expressed",
          column.title = "Samples")
```


###  Volcano plot

After process the differential gene expression analysis (Ex. running the `run_DE_analysis.pl` from [Trinity framework](https://github.com/trinityrnaseq/trinityrnaseq/wiki/Trinity-Differential-Expression)) we can improve the data visualization as follow:

```{r DE, echo = TRUE}
DE <- read.table(paste0(dir,"RSEM.isoform.counts.matrix.Female_vs_Undiff.edgeR.DE_results"))

```


```{r p, fig.cap = "This is a Volcano plot", fig.align = "center", echo = TRUE}
library(ggplot2)
library(scales)

p <- ggplot(DE, aes(x=logFC, y=-log10(PValue))) + geom_point()
p
```

Let's label by Fold Change (up/down) rate and significancy by the follow condition:

* log Fold Change (logFC) > abs(2) 
* False Discovery Rate (FDR) < 0.05 
* Both statements: FDR < 0.05 and logFC > abs(2) 


```{r, fig.cap = "Volcano; Color and fill", fig.align = "center"}
# logf>abs(2) fdr < 0.05 fdr < 0.05 and logfc> abs(2)
DE$Sig <- "Non Sig or basal"
DE$Sig[(abs(DE$logFC) > 2)] <- "logFC"
DE$Sig[(DE$FDR<0.05) & (abs(DE$logFC)>2)] <- "logFC_FDR"

p <- ggplot(DE, aes(x=logFC, y=-log10(PValue))) +
    geom_point(aes(color = Sig)) + theme_classic() +
    # RColorBrewer::display.brewer.all()
    scale_color_brewer()
p
```

Add lines and axis name:

```{r p1, echp = FALSE}
p1 <- p +
        geom_vline(xintercept = 0) +
        geom_hline(yintercept = 0) +
        geom_hline(yintercept = -log10(0.0001), linetype = "dashed") +
        geom_vline(xintercept = c(-2, 2), linetype = "dashed")
```

And also let's rename x and y axis using backquote macros:

```{r, fig.cap = "Volcano lines and axis labeled", fig.align = "center" }
p1 + xlab(bquote(~Log[2]~ "fold change")) + ylab(bquote(~-Log[10]~italic(P)))
# p2 <- p + xlab("Fold change (log2)") + ylab("P-Value")
```

Finally, let's label the dots from the scatter:

```{r p2, fig.cap = "Volcano labels", fig.align = "center"}
library(ggrepel)
maxlab <- max(-log10(DE$PValue)) - 1 # select the points below the highest -log10(PValue) value to label

p2 <-  p1 + geom_text_repel(
          data = subset(DE, -log10(PValue) > maxlab),
          aes(label = Sig),
          size = 2.5,
          box.padding = unit(0.2, "lines"),
          point.padding = unit(0.2, "lines") 
  )
```

```{r, echo = FALSE}
# ref https://github.com/tidyverse/ggplot2/wiki/Share-a-legend-between-two-ggplot2-graphs
grid_arrange_shared_legend <- function(..., ncol = length(list(...)), nrow = 1, position = c("bottom", "right")) {

  plots <- list(...)
  position <- match.arg(position)
  g <- ggplotGrob(plots[[1]] + theme(legend.position = position))$grobs
  legend <- g[[which(sapply(g, function(x) x$name) == "guide-box")]]
  lheight <- sum(legend$height)
  lwidth <- sum(legend$width)
  gl <- lapply(plots, function(x) x + theme(legend.position="none"))
  gl <- c(gl, ncol = ncol, nrow = nrow)

  combined <- switch(position,
                     "bottom" = arrangeGrob(do.call(arrangeGrob, gl),
                                            legend,
                                            ncol = 1,
                                            heights = unit.c(unit(1, "npc") - lheight, lheight)),
                     "right" = arrangeGrob(do.call(arrangeGrob, gl),
                                           legend,
                                           ncol = 2,
                                           widths = unit.c(unit(1, "npc") - lwidth, lwidth)))

  grid.newpage()
  grid.draw(combined)

  # return gtable invisibly
  invisible(combined)

}
```

Finally, let's compare the layouts:

```{r,  fig.cap = "Volcano comparison", fig.align = "center", out.width='\\textwidth', fig.width = 9, fig.height = 5}
library(gridExtra)
library(grid)
grid_arrange_shared_legend(p,p1,p2, ncol = 3)
```


## Functional and enrichment annotation

After assembly denovo or guided transcriptome is common to map each contig to a reference in order to annotate the potential source of each transcript. In this view, blast2go is the average tool used by users to perform this analysis nevertheles blast2go is a non free tool. In spite of, [Trinotate](https://github.com/Trinotate/Trinotate.github.io/wiki) is a useful free-framework to this purpose. Trinotate makes use of a number of different well referenced methods for functional annotation including homology search to known sequence data (BLAST+/SwissProt), protein domain identification (HMMER/PFAM), protein signal peptide and transmembrane domain prediction (signalP/tmHMM), and leveraging various annotation databases (eggNOG/GO/Kegg databases).

In personal experience, trinotate had resulted in better performance (such less computational time) than the blast2go counterpart and useful to automatically annotate one workflow at time.

In addition, [trinotateR](https://github.com/cstubben/trinotateR) is package developed by **Chris Stubben** with useful functions to "wrangling" the tab-delimited `Trinotate.xls` result.

```{r y, echo=TRUE}
library(trinotateR)
y <- read_trinotate(paste0(dir,"Trinotate.xls"))
knitr::kable(summary_trinotate(y))

```

Most of the annotations contain mutliple hits in a backtick-delimited list and each hit contains multiple fields in a caret-delimited list. For example, the second Pfam annotation below contains two hits and each hit contains a pfam id, symbol, name, alignment and e-value. The `split_pfam` functions splits multiple hits and fields, so the second Pfam annotation is now printed in rows 2 and 3 below.


```{r y1, echo =TRUE}
y1 <- split_pfam(y)
head(y1,3)[,c(2,4:7)]
```

Finally, the `summary_pfam` lists both, the number of unique Pfam identifiers and the total  genes, transcripts and proteins with a Pfam annotation.

```{r y2, echo = TRUE}
y2 <- summary_pfam(y1)
knitr::kable(head(y2[order(-y2$transcripts),]))
```

The summary table also includes a count attribute with the number of unique genes, transcripts and proteins with a Pfam annotation, as well as the total number of annotations. In this example, there are 33,721 unique transcripts with a Pfam annotation and 56,642 total annotations to transcripts (since those may have more than one Pfam annotation).

```{r c, echo = TRUE}
c <- attr(y2, "count")
print(c)
```

Implement [*datatable*](https://rstudio.github.io/DT/) to present interactive data-table: from all the available genes/transcripts annotations.

```{r z, echo = FALSE}

z <- data.frame(y2)
z$pfam <- paste0('<a href="http://pfam.xfam.org/family/', z$pfam, '">', z$pfam, '</a>')

library(DT)
datatable(z , escape=1, options = list( pageLength = 10 ) )
#renderDataTable(z , escape=1, options = list( pageLength = 10 ), server = TRUE )
```

## Using `trinotateR` to get the Gene Ontology annotation. 

A GO annotation is a statement about the function of a particular gene. Each GO annotation consists of an association between a gene and a GO term. Together, these statements comprise a “snapshot” of current biological knowledge. The GO describes function with respect to three aspects: molecular function (molecular-level activities performed by gene products), cellular component (the locations relative to cellular structures in which a gene product performs a function), and biological process (the larger processes, or ‘biological programs’ accomplished by multiple molecular activities) 

> Reference: http://www.geneontology.org/page/ontology-documentation, 2018


```{r go, echo = TRUE}
go <- split_GO(y)
gos <- summary_GO(go)
head(gos[order(-gos$transcripts),])
```

And also determine the counts
```{r cgo, echo = TRUE}
cgo <- attr(gos, "count")
print(cgo)
```

Using only the translated transcripts to get the differential Expressed annotations:
```{r got, echo = TRUE}
got <- na.omit(go, cols = "protein")
x$transcript <- rownames(x)
m <- merge(got, x, suffix = c("transcript"), all=FALSE)
m <- m[order(m$transcript),c(1,4,5:12)]

library(DT)
datatable(m , escape=1, options = list( pageLength = 10 ) )

```



Let's draw some visualizations from the annotation enrichment throught a the transcriptome assembly:

```{r plotgos, out.width='\\textwidth', fig.width = 7, fig.height = 7, fig.cap = "Gene Ontology enrichment plot (top 80 transcripts)", fig.align = "center"}

library(ggpubr)
plotgos <- head(gos[order(-gos$transcripts),], 80)
ggbarplot(plotgos, "name", "transcripts",
          fill = "ontology", 
          color = "ontology",
   palette = c("#00AFBB", "#E7B800", "#FC4E07")) +
   theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 7))
   # facet_grid(ontology ~ .,) + theme(#strip.text.x = element_blank(), 
   #                              axis.text.x = element_text(angle = 90, hjust = 1))

```



And separate the ontology annotation for further analysis:

```{r, echo = FALSE}
knitr::kable(table(gos$ontology))
```

You can separete the go terms to perfom further test:

```{r}
MF <- gos[gos$ontology=="molecular_function",]
CC <- gos[gos$ontology=="cellular_component",]
BP <- gos[gos$ontology=="biological_process",]
```

## Semantic similarity (of GO terms)

Determine the similarity of two GO terms based on the annotation statistics of their common ancestor terms by computing semantic similarity among GO terms, sets of GO terms, gene products, and gene clusters, providing different methods than **measure** the information content (IC). To details please read the [Wang's](https://academic.oup.com/bioinformatics/article/26/7/976/213143) paper published in Oxford.


First, build annotation data needed by GOSemSim via `godata` function. Based in figure from Gene Ontology Enrichment, we could focus on the Cellular component (CC) terms.

```{r}
library(org.Hs.eg.db)
hsGO <- GOSemSim::godata('org.Hs.eg.db', ont="CC", computeIC=FALSE)
go <- as.vector(CC$go)
go1 <- sample(go, 20)
go2 <- sample(go, 20)
gosim <- GOSemSim::mgoSim(go1, go2, semData=hsGO, measure="Wang", combine=NULL)
```

And visualize the similarity of the GO term.

```{r, out.width='\\textwidth', fig.width = 7, fig.height = 7}
superheat(gosim,
          # retain original order of rows/cols
          pretty.order.rows = TRUE,
          pretty.order.cols = TRUE,
          #left.label = "none",
          bottom.label.text.angle = 90,
          row.title = "Sample 1",
          column.title = "Sample 2",
          bottom.label.text.size = 4,
          left.label.text.size = 4
          )
```

Finally detache (unload) `org.Hs.eg.db` package.

```{r, echo = TRUE}
detach(package:org.Hs.eg.db, unload = TRUE)
```

<!-- # bottom.label = "none" -->
<!-- gosim2 <- as.data.frame(gosim) -->
<!-- db <- rbind(CC[CC$go %in% rownames(gosim2), c(1,3)], CC[CC$go %in% colnames(gosim2), c(1,3)]) -->

## Blast hits annotation

From the annotation, lets use the blastx result in order to get the lineage from proteins anntotated.

```{r, echo = FALSE}
split_blast2 <- function (x, hit = "sprot_Top_BLASTX_hit") 
{
  y <- x[!is.na(get(hit)), .(get(hit), gene_id, transcript_id, 
                             prot_id)]
  z <- strsplit(y$V1, "`")
  n <- sapply(z, length)
  z <- strsplit(unlist(z), "\\^")
  if (any(sapply(z, "[", 1) != sapply(z, "[", 2))) 
    print("WARNING: check different values in columns 1 and 2")
  NAME <- gsub("^RecName: Full=", "", sapply(z, "[", 6))
  NAME <- gsub("SubName: Full=", "", NAME)
  NAME <- gsub(";$", "", NAME)
  NAME <- gsub(" \\{[^}]+}", "", NAME)
  x1 <- data.frame(gene = rep(y$gene_id, n), transcript = rep(y$transcript_id, 
                                                              n), protein = rep(gsub(".*\\|", "", y$prot_id), n), uniprot = sapply(z, 
                                                                                                                                   "[", 1), align = sapply(z, "[", 3), identity = as.numeric(gsub("%ID", 
                                                                                                                                                                                                  "", sapply(z, "[", 4))), evalue = as.numeric(gsub("E:", 
                                                                                                                                                                                                                                                    "", sapply(z, "[", 5))), name = NAME, lineage = sapply(z, 
                                                                                                                                                                                                                                                                                                           "[", 7), domain = gsub("; .*", "", sapply(z, "[", 7)), 
                   genus = gsub(".*; ", "", sapply(z, "[", 7)), stringsAsFactors = FALSE)
  message(nrow(x1), " ", hit, " annotations")
  data.table(x1)
}
```



```{r, echo = TRUE}
blast <- split_blast2(y)
blast2 <- summary_blast(blast)
knitr::kable(head(blast2[order(-blast2$transcripts),]))
```


## Taxonomic tree:
Visualizing hierarchical information 
### Metacoder [*Metacoder*](https://grunwaldlab.github.io/metacoder_documentation/vignettes--02--plotting_taxonomic_data.html). 

Let's subset the virus lineage annotation:

```{r}
vs <- blast[blast$domain == "Viruses",]
list <- vs$lineage
```

And plot ...

```{r obj, out.width='\\textwidth', fig.width = 7, fig.height = 7, fig.align = "center", fig.cap = "Tree example using metacoder", echo = TRUE}

library(metacoder)
library(RColorBrewer)
obj <- parse_tax_data(vs, class_cols = "lineage", class_sep = ";")
display.brewer.pal(n = 5, name = 'Set3') # Hexadecimal color specification
heat_tree(obj, 
          node_label = taxon_names,
          node_size = n_obs,
          node_color = n_obs,
          #edge_color = "grey",
          node_color_range = brewer.pal(n = 10, name = "Oranges"))
```

Which annotation could be true, let's show the identity ?

```{r, out.width='\\textwidth', fig.width = 5, fig.height = 5, fig.align = "center", fig.cap = "Tree example 2; labeling by identity", echo = TRUE}
heat_tree(obj, 
          edge_label = taxon_names,
          edge_label_size = 0.1,
          node_color = identity,
          edge_color = "grey",
          node_color_range = brewer.pal(n = 10, name = "Oranges"),
          node_size_axis_label = "n_obs",
          node_color_axis_label = "Identity")
```

### [*ggtree*]()
The ggtree is designed by extending the ggplot2 (Wickham 2009) package. It is based on the grammar of graphics and takes all the good parts of ggplot2. *ggtree* supports several layouts, including rectangular, slanted, circular and fan for phylogram and cladogram, equal_angle and daylight for unrooted layout, time-scaled and two dimentional phylogenies.

> Reference: Yu et al. 2017, https://besjournals.onlinelibrary.wiley.com/doi/abs/10.1111/2041-210X.12628

```{r, echo = TRUE}
tbl <- head(vs[order(-vs$identity),], 50)
tbl <- tbl[!duplicated(tbl$transcript),] # remove duplicates annotations
write.table(tbl$transcript, file = "virus.list", quote = FALSE, col.names = FALSE, row.names = FALSE)
```


<!-- python ~/Documents/GitHub/IIIBSS/Final_Project/code/get.seqs.py ./infile/virus.list ./Trinity.fasta virus.fasta && awk '/^>/{print $1; next}{print}' < virus.fasta > virus.fst -->

Run in clustalo web-service and input again in r the tree
this is a Neighbour-joining tree without distance corrections.

```{r, echo = TRUE}
tree <- treeio::read.newick(paste0(dir,"virus.tree.corrected.txt"))
```

And plot

```{r, echo = TRUE, out.width='\\textwidth', fig.width = 10, fig.height = 7, fig.align = "center", fig.cap = "ggtree; using different layouts"}
library(ggtree)
multiplot(
        ggtree(tree, branch.length = "none") + geom_treescale(width=0.4),
        ggtree(tree, branch.length='none', layout="daylight") + geom_treescale(width=0.4) + geom_nodepoint(),
        ggtree(tree, branch.length='none', layout='circular') + geom_treescale(width=0.4),
        ncol=3,labels = LETTERS[1:3])

```
Then

```{r, echo = TRUE}
p <-  ggtree(tree, branch.length='none', layout='circular') + geom_treescale(width=0.4) +
      geom_nodepoint(color="#b5e521", alpha=1/4, size=4) 
```

Modify tips and node geometry

```{r, echo = TRUE, out.width='\\textwidth', fig.width = 10, fig.height = 7, fig.align = "center", fig.cap = "ggtree; using different layouts"}
multiplot(# add node points
        p + geom_nodepoint(),
        # add tip points
        p + geom_tippoint(),
        # Label the tips
        p + geom_tiplab(),
        ncol=3,labels = LETTERS[1:3])
```

Use annotation info to merge within object (ie. tree and dataframe)

```{r, echo = TRUE, out.width='\\textwidth', fig.width = 9, fig.height = 6, fig.align = "center", fig.cap = "ggtree; labeling tree based on data-frame" }
d <- p + geom_nodepoint()

tbl <- tbl[,-c(1)]
colnames(tbl)[1] <- "label"
tbl$genus <- gsub("unclassified ", "", tbl$genus)


t1 <- d %<+% tbl + 
    geom_point(aes(color=genus), size=5, alpha=.5, na.rm = TRUE) +  theme(legend.position="bottom")

t2 <- p %<+% tbl + 
    geom_tiplab(size=2.5, aes(label=paste0('italic(', uniprot, ')')), parse=TRUE)

multiplot(t1, t2, ncol=2)

```


## Final words
Just enjoy the course!



<!-- ``` {sh}    -->
<!-- after running let docker the 	`simple_phylogeny_lwp.pl` -->

<!-- docker run --rm -it ebiwp/webservice-clients simple_phylogeny_lwp.pl --help -->

<!-- docker run --rm -it ebiwp/webservice-clients simple_phylogeny_lwp.pl --kimura on --email rgomez@cicese.edu.mx  virus.fsa.clustal_num -->
<!-- ``` -->

<!-- Melting files -->
<!-- # ```{r echo = FALSE} -->
<!-- # x <- melt(z, id.vars = c("Temperature", "Condition"), -->
<!-- #           #measure.vars = c("Hif1α", "Hsp70", "Hsp90", "GPx", "cMnSOD", "Na-K ATPasa"), -->
<!-- #           value.name= "expression", variable_name = "Name") -->
<!-- #  -->
<!-- # colnames(x)[c(3:4)] <- c("Gene", "Expression") -->
